test/integration_tests/test_lanchain_llm.py

```diff
@@ -0,0 +1,92 @@
+# coding=utf-8
+# Copyright 2018-2023 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import unittest
+from test.util import (
+    load_udfs_for_testing,
+    shutdown_ray,
+    suffix_pytest_xdist_worker_id_to_dir,
+)
+
+from evadb.configuration.constants import EVA_DATABASE_DIR, EVA_ROOT_DIR
+from evadb.interfaces.relational.db import connect
+
+
+class TestLangchainLLM(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def setUpClass(cls):
+        cls.db_dir = suffix_pytest_xdist_worker_id_to_dir(EVA_DATABASE_DIR)
+        cls.conn = connect(cls.db_dir)
+        cls.evadb = cls.conn._evadb
+
+    def setUp(self):
+        self.evadb.catalog().reset()
+        load_udfs_for_testing(
+            self.evadb,
+        )
+
+    def tearDown(self):
+        shutdown_ray()
+
+    def test_langchain_llm_gpt4all(self):
+        conn = connect()
+        cursor = conn.cursor()
+        # loading a pdf in eva
+        pdf_path1 = f"{EVA_ROOT_DIR}/data/documents/state_of_the_union.pdf"
+
+        load_pdf = cursor.load(file_regex=pdf_path1, format="PDF", table_name="PDFss")
+        load_pdf.execute()
+        # droping SentenceTransformerFeatureExtractor udf if existes
+        cursor.drop_udf("SentenceTransformerFeatureExtractor", if_exists=True).execute()
+        # creating SentenceTransformerFeatureExtractor udf
+        udf = cursor.create_udf(
+            "SentenceTransformerFeatureExtractor",
+            True,
+            f"{EVA_ROOT_DIR}/evadb/udfs/sentence_transformer_feature_extractor.py",
+        )
+        udf.execute()
+        # droping GPT4AllQaUDF udf if existes
+        cursor.drop_udf("GPT4AllQaUDF", if_exists=True).execute()
+        # creating GPT4AllQaUDF udf
+        udf = cursor.create_udf(
+            "GPT4AllQaUDF",
+            True,
+            f"{EVA_ROOT_DIR}/evadb/udfs/GPT4ALL.py",
+        )
+        udf.execute()
+        # fetching Similarity based on features from SentenceTransformerFeatureExtractor udf and returning top 3 results
+        pdf_table_similarity = (
+            cursor.table("PDFss")
+            .order(
+                """Similarity(
```

Comment:
I am a bit confused here. If we are already ranking the source similarity by using sentence transformer. Why do we need another index and embedding inside the GPT4All UDF?

Reply:
we are not creating index two times for sentence transformer and in chat4gptall udf , we are just creating the index in chat4gptall all , as you can see in the udf and test it must be clear @jiashenC

Reply:
1. My confusion is why you need this similarity search outside UDF? You do have an FAISS index inside UDF, which will be doing similarity search when you pass the question and source to it.
2. More of a question to @gaurav274, do you want to keep an index inside UDF here, which is transparent to the system.

Reply:
so , first we are doing similarity for filtering and pass only the top 3 or 5 results for answers to the gpt4allqa udf